content/2017.02.handling-python-projects-dependencies.rst

```diff
@@ -0,0 +1,392 @@
+
+What is happening is that Python is checking the dependencies and will
+refuse to run if you have a greater ``requests`` version.
+
+This can happen if another dependency already needed ``requests`` as a
+dependency and pip already installed it with a greater version.
+
+**So really do it only if you must.**
+
+
+What about test dependencies?
+=============================
+
+That's a good question, I am glad you asked.
+
+Test dependencies are less of an issue. You can either use
+``test_requires`` in your ``setup.py`` or a ``dev-requirements.txt``
+file.
+
+In the former you will need to run tests using ``python setup.py test``
+which unfortunately doesn't install dependencies in a ``virtualenv``
+
+In the later you will need to make sure your test dependencies are
+installed before running the tests but tools like ``tox`` already do
+that for you.
+
+Our take on this was to put test dependencies in a
+``dev-requirements.txt`` file.
+
+We have a Makefile rule that knows if dev-dependencies should be
+installed or not before running the ``tests`` target.
+
+As a bonus it will automatically create a virtualenv if you don't
+have one already activated:
+
+.. code-block:: Makefile
+
+    VIRTUALENV = virtualenv --python=python3
+    VENV := $(shell echo $${VIRTUAL_ENV-.venv})  # Use the activated virtualenv path or use .venv
+    PYTHON = $(VENV)/bin/python
+    INSTALL_STAMP = $(VENV)/.install.stamp
+    INSTALL_DEV_STAMP = $(VENV)/.dev_env_installed.stamp
+    
+    install: $(INSTALL_STAMP)
+    
+    $(INSTALL_STAMP): $(PYTHON) setup.py  # Refresh the virtualenv if setup.py changed
+    	$(VENV)/bin/pip install -U pip
+    	$(VENV)/bin/pip install -Ue .
+    	touch $(INSTALL_STAMP)
+    
+    virtualenv: $(PYTHON)  # Create the virtualenv if needed (python executable not present)
+    
+    $(PYTHON):
+    	$(VIRTUALENV) $(VENV)
+    	
+    install-dev: install $(DEV_STAMP)
+    
+    # Refresh dev dependencies if dev-requirements.txt changed
+    $(DEV_STAMP): $(PYTHON) dev-requirements.txt
+    	$(VENV)/bin/pip install -Ur dev-requirements.txt
+    	touch $(DEV_STAMP)
+    
+    tests-once: install-dev
+        $(VENV)/bin/py.test tests/
+    
+    serve: install
+    	$(VENV)/bin/python manage.py runserver
+
+
+Conclusion
+==========
+
+As a conclusion, when working with Python dependencies there are three
+process that needs to work well together.
+
+- Installing automatically a project version in a working state,
+  that's what you can do using requiremrents constraint files.
+- Keeping the project up-to-date with next version of its dependencies
+  without breaking previous released versions.
+- Installing a project as part as another project without it breaking
+  the project with dependencies version conflicts.
+
+Our solution to this is as follow:
+
+- Keep your project dependencies clean of any version number.
+- At release time document what is the known version check.
```

Comment:
I don't like `known version check`. I guess this is `known good version`, or `tested dependency versions` or something. Also, nit: a comma after `At release time` would be good.